### PR TITLE
feature/DOS-2289: Added ability to distribute units across AZs.

### PR DESCRIPTION
**Pull request by Grindizer**
Added ability to distribute units across AZs.
Fleet units of application will be distributed across the list of availability zones defined in 
/deis/controller/AZs, if the list does not exists nothing change, and the normal behavior is observed.

### CODE REVIEW

File: controller/scheduler/fleet.py [hunk 1]
```diff
@@ -115,6 +115,15 @@ def _get_machines(self):
         return json.loads(data)
 
     # container api
+    def _get_az_for(self, name):
+        num = int(re.match(MATCH, name).groupdict()['c_num'])
+        az_tag = {}
```
Comment (DanHoerst) [inline controller/scheduler/fleet.py]: You don't need to instantiate this here if you are using `dict( )` below.

File: controller/scheduler/fleet.py [hunk 2]
```diff
@@ -115,6 +115,15 @@ def _get_machines(self):
...
+        AZs = settings.AVAILABILITY_ZONE_LIST
+        if AZs:
+            index = num % len(AZs)
+            az_tag = dict(az=AZs[index])
```
Comment (aj-may) [inline controller/scheduler/fleet.py]: This can be set like this:

```
az_tag = {
    'az': AZs[index],
}
```

File: controller/scheduler/fleet.py [hunk 3]
```diff
@@ -115,6 +115,15 @@ def _get_machines(self):
         return json.loads(data)
 
     # container api
+    def _get_az_for(self, name):
+        num = int(re.match(MATCH, name).groupdict()['c_num'])
```
Comment (aj-may) [inline controller/scheduler/fleet.py]: This variable should be more descriptive.

maybe `container_number`?

File: controller/scheduler/fleet.py [hunk 4]
```diff
@@ -124,6 +133,7 @@ def create(self, name, image, command='', template=None, **kwargs):
     def _create_container(self, name, image, command, unit, **kwargs):
         l = locals().copy()
         l.update(re.match(MATCH, name).groupdict())
+        azs = self._get_az_for(name)
```
Comment (aj-may) [inline controller/scheduler/fleet.py]: why don't we run this by the related logic around line 172

File: controller/scheduler/fleet.py (hunk 1 above)
Comment (DanHoerst) [inline controller/scheduler/fleet.py:120]: It's probably more clear to put this in the `else` of `if AZs:`

File: controller/templates/confd_settings.py [hunk 5]
```diff
@@ -79,3 +79,12 @@
...
 {{ end }}
+
+# if defined, then scheduler will distribute the units cross that list.
+# it is a list of anything, the scheduler would pick az = AVAILABILITY_ZONE_LIST[num % len(AVAILABILITY_ZONE_LIST)]
+# which means if list is ['a', 'b'], and it is the 3rd container of the same application,
+# we would do, 3 % 2 = 1 => az = list(1) => az=b
+AVAILABILITY_ZONE_LIST=[]
+{{ if exists "/deis/controller/AZs" }}
+AVAILABILITY_ZONE_LIST=json.loads('''{{ getv "/deis/controller/AZs" }}'''.replace("\'", '"'))
+{{ end }}
```
Comment (DanHoerst) [inline controller/templates/confd_settings.py]: Can you please add this new line.

File: controller/scheduler/fleet.py [hunk 6]
```diff
@@ -158,6 +168,10 @@ def _create_container(self, name, image, command, unit, **kwargs):
             f['value'] = f['value'].format(**l)
         # prepare tags only if one was provided
         tags = kwargs.get('tags', {})
+        # this way, if we specify the az tag it will be higher priority then the
+        # cross az feature.
+        azs.update(tags)
```
Comment (aj-may) [inline controller/scheduler/fleet.py:175]: can we change these 2 lines to:

```
tags = self._get_az_for(name).update(tags)
```

File: controller/templates/confd_settings.py [hunk 7]
```diff
@@ -79,3 +79,12 @@
 PRIVATE_EXT_REGISTRY_FLAVOR = '{{ getv "/deis/controller/private_ext_registry/flavor" }}'
 PRIVATE_EXT_REGISTRY_DATA = json.loads('''{{ getv "/deis/controller/private_ext_registry/data" }}'''.replace("\'", '"'))
 {{ end }}
+
+# if defined, then scheduler will distribute the units cross that list.
+# it is a list of anything, the scheduler would pick az = AVAILABILITY_ZONE_LIST[num % len(AVAILABILITY_ZONE_LIST)]
+# which means if list is ['a', 'b'], and it is the 3rd container of the same application,
+# we would do, 3 % 2 = 1 => az = list(1) => az=b
+AVAILABILITY_ZONE_LIST=[]
```
Comment (DanHoerst) [inline controller/templates/confd_settings.py]: I think this can also go in the else of `{{ if exists "/deis/controller/AZs" }}`

File: controller/scheduler/fleet.py (hunk 6 above)
Comment (Grindizer) [inline controller/scheduler/fleet.py:175]: actually, we can't, the update function in python updates the dict in place, it does not return the dict. 
So the result of your command would be tags = None.

File: controller/templates/confd_settings.py (hunk 5 above)
Comment (Grindizer) [inline controller/templates/confd_settings.py]: Done.

File: controller/templates/confd_settings.py (hunk 7 above)
Comment (Grindizer) [inline controller/templates/confd_settings.py]: Honestly, i'd rather not, this is not an unknown pattern in python, and I wanted to stay coherent with the lines above where we use the same style. 
Same thing above in fleet.py, the same pattern is used by deis devs you can see it just above, again, I just tried to stay coherent. 

```
data = resp.read()
if resp.status not in (200,):
    errmsg = "Failed to retrieve machines: {} {} - {}".format(
                resp.status, resp.reason, data)
    raise RuntimeError(errmsg)
return json.loads(data)
```

Comment (DanHoerst) [inline controller/templates/confd_settings.py]: I don't think what you've posted is related to instantiating an empty list/dict. Line 8 of this file has a good example of what I'm recommending:

```
try:
    SCHEDULER_OPTIONS = dict('{{ if exists "/deis/controller/schedulerOptions" }}{{ getv "/deis/controller/schedulerOptions" }}{{ else }}{}{{ end }}')
except:
    SCHEDULER_OPTIONS = {}
```

Comment (Grindizer) [inline controller/templates/confd_settings.py]: What I posted was to show the pattern 

```
value = -init-
if ...: 
     use/modify value
return value
```

related to you comment above about azs.

For here, I said I want to stay coherent with the line just above it. 
We can discuss coding style if you want, and if we agree on something I'd be happy to raise another PR that addresses all the necessary changes, but this is out of the scope of this PR.

File: controller/templates/confd_settings.py [hunk 8]
```diff
@@ -79,3 +79,13 @@
 PRIVATE_EXT_REGISTRY_FLAVOR = '{{ getv "/deis/controller/private_ext_registry/flavor" }}'
 PRIVATE_EXT_REGISTRY_DATA = json.loads('''{{ getv "/deis/controller/private_ext_registry/data" }}'''.replace("\'", '"'))
 {{ end }}
+
+# if defined, then scheduler will distribute the units cross that list.
+# it is a list of anything, the scheduler would pick az = AVAILABILITY_ZONE_LIST[num % len(AVAILABILITY_ZONE_LIST)]
+# which means if list is ['a', 'b'], and it is the 3rd container of the same application,
+# we would do, 3 % 2 = 1 => az = list(1) => az=b
+{{ if exists "/deis/controller/AZs" }}
```
Comment (aj-may) [inline controller/templates/confd_settings.py]: in the deis codebase it follows this pattern:

```
GROUP_TYPE = '{{ if exists "/deis/controller/auth/ldap/group/type" }}{{ getv "/deis/controller/auth/ldap/group/type"}}{{ else }} {{ end }}'
```

Lets keep consistent with what they do

Comment (Grindizer) [inline controller/templates/confd_settings.py]: It is not exactly the same thing. I update 3 value is private_ext_registry exists. 
If I do it in one line, I would need to repeat 
`{{ if exists "/deis/controller/private_ext_registry/flavor" }}` 
for every value, which we don't want because we want it to fail flavor is set and data is not.

Comment (aj-may) [inline controller/templates/confd_settings.py]: are you talking about this line?

```
AVAILABILITY_ZONE_LIST=json.loads('''{{ getv "/deis/controller/AZs" }}'''.replace("\'", '"'))
```

this looks like one value

Comment (Grindizer) [inline controller/templates/confd_settings.py]: No the one before, for that one I updated as Dan suggested.
